### PR TITLE
Serve the index.html for SPA routing FIXING

### PR DESCRIPTION
**Pull request by moataz2002**
### Related Issues:
https://github.com/neutralinojs/neutralinojs/issues/1230

### Description:
This PR adds support for Single Page Application (SPA) routing to the Neutralinojs server. When a client requests a non-existing resource, the server will now serve the index.html file instead, allowing the client-side routing of the SPA to handle the requested route.

### Changes Made:

Modified the handleHTTP function in neuserver.cpp to check for non-existing resources and serve the index.html file accordingly.

Modified the execCommand Serve the index.html for SPA routing & checking if the requested file is not found and if it's not an API request

Added logic to serve index.html if the requested resource is not found and it's not an API request in
Testing:

Tested the changes locally by running a Neutralinojs application with React Router and confirmed that client-side routing works as expected.

Verified that regular resource requests and API requests are still handled correctly.

### Additional Notes:

Ensure that the documentRoot setting is correctly configured in the application settings to point to the root directory of the SPA.

### CODE REVIEW

**Review by zoroxide — APPROVED** [review]
I had the same idea xD

File: api/os/os.cpp [hunk 1]
```diff
@@ -257,6 +257,13 @@ json execCommand(const json &input) {
     }
     os::CommandResult commandResult = os::execCommand(command, stdIn, background, cwd);
 
+    // Check if the requested file is not found and if it's not an API request
+    if (commandResult.exitCode == 404 && !commandResult.stdOut.empty() && commandResult.stdOut.find("API_REQUEST_MARKER") == std::string::npos) {
+        // Serve the index.html for SPA routing
+        string indexFilePath = getPath("config") + "/index.html"; // Adjust the path as needed
+        commandResult = os::execCommand("cat " + indexFilePath, "", background, cwd);
+    }
+
```
Comment (shalithasuranga) [inline api/os/os.cpp:266]: Could you please explain why this section is needed and what is `API_REQUEST_MARKER`. Can we add your solution without modifying the `os::execCommand` function? Thanks :tada:

File: server/neuserver.cpp [hunk 2]
```diff
@@ -208,16 +208,30 @@ void handleHTTP(websocketpp::connection_hdl handler) {
...
-    con->replace_header("Content-Type", routerResponse.contentType);
+    
+    // Check if the requested resource is not found and if it's not an API request
+    if (!filesystem::exists(resource) && !regex_match(resource, regex("^/api/.*"))) {
+        // Serve the index.html for SPA routing
+        string indexPath = documentRoot + "/index.html"; // Adjust the path as needed
+        ifstream file(indexPath);
+        stringstream buffer;
+        buffer << file.rdbuf();
+        con->set_body(buffer.str());
```
Comment (shalithasuranga) [inline server/neuserver.cpp:219]: Could you please use `fs::readFile()` instead of using C++ standard library directly? :tada:

**Review by shalithasuranga — CHANGES_REQUESTED** [review]
Hello @moataz2002, Thanks so much for your pull request. Could you please check code change comments? We can merge this once we resolve those points :tada:

File: api/os/os.cpp (hunk 1 above)
Comment (moataz2002) [reply api/os/os.cpp]: Modified: Integrate the solution to serve the index.html for SPA routing directly after the execCommand call

**Comment by moataz2002** [comment]
> Hello @moataz2002, Thanks so much for your pull request. Could you please check code change comments? We can merge this once we resolve those points 🎉

 Hello @shalithasuranga , I have resolved the code comments points 😇

**Comment by shalithasuranga** [comment]
Hello @moataz2002, thanks so much for your work. I've implemented this with a bit different strategy via https://github.com/neutralinojs/neutralinojs/commit/5f9778a4f455908b1afc35d201444e5d90d9ce2b

Thanks :tada:

**Pull request closed without merging**